velox/functions/lib/SIMDComparisonUtil.h

```diff
@@ -233,70 +239,189 @@ void applyAutoSimdComparison(
+          ->valueVector()
+          ->isFlatEncoding()) {
+    auto dictB = args[1]->asUnchecked<DictionaryVector<B>>();
+    const A* __restrict rawA =
+        args[0]->asUnchecked<FlatVector<A>>()->rawValues();
+    const B* __restrict rawB = dictB->valueVector()
+                                   ->template asUnchecked<FlatVector<B>>()
+                                   ->rawValues();
+    const vector_size_t* __restrict indexB =
+        dictB->indices()->template as<vector_size_t>();
+    detail::applyAutoSimdComparisonInternal(
+        [&](const A* __restrict rawA,
+            const B* __restrict rawB,
+            const vector_size_t* __restrict indexA,
+            const vector_size_t* __restrict indexB,
+            int i) {
           if constexpr (sizeof...(cmpArgs) > 0) {
-            return Compare::apply(rawA[0], rawB[0], cmpArgs...);
+            return Compare::apply(rawA[i], rawB[indexB[i]], cmpArgs...);
           } else {
-            return cmp(rawA[0], rawB[0]);
+            return cmp(rawA[i], rawB[indexB[i]]);
+          }
+        },
+        result,
+        rows,
+        rawA,
+        rawB,
+        nullptr,
+        indexB);
+  } else if (
+      args[1]->isFlatEncoding() &&
+      args[0]->encoding() == VectorEncoding::Simple::DICTIONARY &&
+      args[0]
+          ->asUnchecked<DictionaryVector<A>>()
+          ->valueVector()
+          ->isFlatEncoding()) {
+    auto dictA = args[0]->asUnchecked<DictionaryVector<A>>();
+    const A* __restrict rawA = dictA->valueVector()
+                                   ->template asUnchecked<FlatVector<A>>()
+                                   ->rawValues();
+    const vector_size_t* __restrict indexA =
+        dictA->indices()->template as<vector_size_t>();
+    const B* __restrict rawB =
+        args[1]->asUnchecked<FlatVector<B>>()->rawValues();
+    detail::applyAutoSimdComparisonInternal(
+        [&](const A* __restrict rawA,
+            const B* __restrict rawB,
+            const vector_size_t* __restrict indexA,
+            const vector_size_t* __restrict indexB,
+            int i) {
+          if constexpr (sizeof...(cmpArgs) > 0) {
+            return Compare::apply(rawA[indexA[i]], rawB[i], cmpArgs...);
+          } else {
+            return cmp(rawA[indexA[i]], rawB[i]);
           }
         },
-        result);
+        result,
+        rows,
+        rawA,
+        rawB,
+        indexA,
+        nullptr);
   } else {
     VELOX_UNREACHABLE();
   }
 }
+
+template <typename A, typename B>
+bool shouldApplyAutoSimdComparison(
+    const SelectivityVector& rows,
+    std::vector<VectorPtr>& args) {
+  if (rows.end() - rows.begin() > 64) {
+    if ((args[0]->isFlatEncoding() || args[0]->isConstantEncoding()) &&
+        (args[1]->isFlatEncoding() || args[1]->isConstantEncoding())) {
+      return true;
+    } else if (
+        args[0]->isFlatEncoding() &&
+        args[1]->encoding() == VectorEncoding::Simple::DICTIONARY &&
```

Comment:
Are you sure we are hitting this code path?  If arguments are one flat and one dictionary, we would peel the dictionary and apply it on the two flat and unpeel the result.

Reply:
@bikramSingh91 Just want to confirm the peeling should be triggered in this?

Reply:
Yes, and we saw the difference with this change. I think the peel doesn't happen is because of https://github.com/facebookincubator/velox/blob/86efb6343e952c90600fd202e6b623e41cd2e050/velox/expression/Expr.cpp#L1050